src/baldr/incident_singleton.h

```diff
@@ -0,0 +1,306 @@
+  // instance. this gives the responsibility to the last living thread to deallocate the state object.
+  // if we didn't do this, and the watcher thread were still running when the singleton instance got
+  // destructed, then the watcher would be making use of a deallocated state object. this way, if the
+  // watcher is last to die it own the lifetime of the state and if the singleton is the last to die
+  // it owns the lifetime of the state. note that we still need to use atomics inside the state as
+  // only the shared_ptr itself is thread safe, not the thing it points to
+  std::shared_ptr<state_t> state;
+  // daemon thread to watch for new incident data
+  std::thread watcher;
+  // the total amount of time we are allowed to make a full update to the incidents cache
+  time_t max_loading_latency;
+
+  /**
+   * Singleton private constructor that static function uses to instantiate the singleton
+   * @param config   lets the daemon thread know where/how to look for incidents
+   * @param tileset  an mmapped graph tileset (ie static) allows incident loading to be lock-free
+   */
+  incident_singleton_t(const boost::property_tree::ptree& config,
+                       const std::unordered_set<valhalla::baldr::GraphId>& tileset)
+      : state{new state_t}, watcher(watch, config, std::cref(tileset), state),
+        max_loading_latency(
+            config.get<time_t>("incident_max_loading_latency", DEFAULT_MAX_LOADING_LATENCY)) {
+    // see if the thread can start up and do a pass to load all the incidents
+    std::unique_lock<std::mutex> lock(state->mutex);
+    auto when = std::chrono::system_clock::now() + std::chrono::seconds(max_loading_latency);
+    if (state->signal.wait_until(lock, when) == std::cv_status::timeout) {
+      throw std::runtime_error("Unable to initialize incident watcher in the configured time period");
+    }
+    // let this run forever
+    watcher.detach();
+  }
+
+  /**
+   * Updates the tile in the states cache
+   * @param state     the state to update
+   * @param tile_id   the tile id we are loading
+   * @param path      the path to the tile file
+   * @return the shared_ptr that owns the tile memory
+   */
+  static bool update_tile(const std::shared_ptr<state_t>& state,
+                          const valhalla::baldr::GraphId& tile_id,
+                          const std::string& path) {
+    // load the tile
+    auto tile = read_tile(path);
+    // see if we have a slot
+    auto found = state->cache.find(tile_id);
+    // if we dont have a slot make one, this must be synchronized
+    if (found == state->cache.cend()) {
+      // this shouldnt happen in lock free mode but can if you put unexpected tiles in the log/dir
+      if (state->lock_free.load()) {
+        LOG_WARN("Incident watcher skipped " + std::to_string(tile_id) +
+                 " because it was not found in the configured tile extract");
+        return false;
+      }
+      state->mutex.lock();
```

Comment:
Perhaps it would be safer to use a `unique_lock` that gets automatically destructured to prevent a deadlock if the hashtable `insert` below throws?

Reply:
great catch! ill use a scoped lock